### PR TITLE
[SPARK-29273][core] Save peakExecutionMemory value when writing task end to event log

### PR DESCRIPTION
**Pull request by 012huang**
in TaskMetrics, there is a exposed metrics peakExecutionMemory, but the value never been set. when a task is finished, it generate a SparkListenerTaskEnd event info, incuding the metrics value. actually the peakExecutionMemory is stored in the Accumulables which is a member of TaskInfo.

so when parse the SparkListenerTaskEnd event, we can get the `internal.metrics.peakExecutionMemory` value from the parsed taskInfo object, and set back to the TaskMetricsInfo with supply a setPeakExecutionMemory method.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 1]
```diff
@@ -647,6 +647,13 @@ private[spark] object JsonProtocol {
     val taskInfo = taskInfoFromJson(json \ "Task Info")
     val executorMetrics = executorMetricsFromJson(json \ "Task Executor Metrics")
     val taskMetrics = taskMetricsFromJson(json \ "Task Metrics")
+    val peakExecutionMemory = taskInfo.accumulables.find(accInfo => {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: This code should be in `taskMetricsFromJson` like other metrics, following the style over there.

Also, please add a unit test.

Comment (012huang) [reply core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: Thanks for your reply, I will add soon

Comment (Tony-huangweiyi) [reply core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: hi, @vanzin , I update this PR

the peakExecutionMemory value zero happend when replaying event log, so I add the metrics value in SparkListenerTaskEnd json info and parsing it when replaying and set the value back like other metrics do, this is a litter more natural than the first commit for this PR

the unit test is the same as JsonProtocolSuite.testTaskMetrics 

could you please kindly review?

Comment (012huang) [reply core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: hi, @vanzin , I update this PR

the peakExecutionMemory value zero happend when replaying event log, so I add the metrics value in SparkListenerTaskEnd json info and parsing it when replaying and set the value back like other metrics do, this is a litter more natural than the first commit for this PR

the unit test is the same as JsonProtocolSuite.testTaskMetrics

could you please kindly review?

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: You need to update that unit tests to verify the new value you're adding.

**Comment by vanzin** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala (hunk 1 above)
Comment (012huang) [reply core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: thanks, I update that unit test and take a successful test before push.

**Comment by vanzin** [comment]
Merging to master,

**Pull request closed without merging**